Added components to dynamically set joint limits

peci1 opened the pull request:
# 🎉 New feature

## Summary
This PR adds support for dynamic setting of joint limits (position, velocity and effort).

Depends on https://github.com/ignitionrobotics/ign-physics/pull/260 (and also https://github.com/ignition-forks/dart/issues/26, though it is not needed if it's no problem that some limits are not enforced).

## Test it
Integration tests were added, see the usage in them.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

Codecheck fails with `src/systems/physics/Physics.cc:1943:  Small and focused functions are preferred: PhysicsPrivate::UpdatePhysics() has 534 non-comment lines (error triggered by exceeding 500 lines).  [readability/fn_size] [1]`, I don't really know what to do about that.

This is aimed at running in DARPA SubT finals.

peci1 commented:
Example usage here: https://github.com/osrf/subt/pull/948 .

azeey commented on include/ignition/gazebo/components/Serialization.hh:
```diff
@@ -183,6 +183,40 @@ namespace serializers
+    public: static std::ostream &Serialize(std::ostream &_out,
+      const std::vector<T> &_data)
+    {
+      _out << _data.size();
```
I think we need to add a whitespace charactor to `_out` after adding `_data.size()` and after adding each `datum`. Currently, if your vector is `[{5,6},{7,8}]`, I think it'll end up being serialized as `25 67 8`.

peci1 replied:
I added the space. I hope the deserialization step doesn't need any change.

azeey commented on src/systems/physics/Physics.cc:
```diff
@@ -1298,6 +1338,99 @@ void PhysicsPrivate::UpdatePhysics(EntityComponentManager &_ecm)
+            _entity);
+        if (posLimits && !posLimits->Data().empty())
+        {
+          auto limits = posLimits->Data();
```
nit: can this be a const ref since `limits` is not going to change?

peci1 replied:
fixed in all 3 places

azeey commented on test/integration/physics_system.cc:
```diff
@@ -812,6 +817,387 @@ TEST_F(PhysicsSystemFixture, ResetVelocityComponent)
+            }
+            else if (iteration == 50u)
+            {
+              EXPECT_NE(nullptr, limitComp);
```
The `limitComp` component may not be valid after creating new components (see https://github.com/ignitionrobotics/ign-gazebo/issues/182), so I recommend fetching the component again here.

azeey replied:
Can you change this to `ASSERT_NE` since we don't want to proceed if the pointer is nullptr?

peci1 replied:
done (both comments)

peci1 replied:
actually, `ASSERT_NE` is not allowed in that place (inside a callback)... so i've kept `EXPECT_EQ` and added a test for the pointer to be non-null

azeey commented on test/integration/physics_system.cc:
```diff
@@ -812,6 +817,387 @@ TEST_F(PhysicsSystemFixture, ResetVelocityComponent)
+            }
+            else if (iteration == 50u)
+            {
+              EXPECT_NE(nullptr, limitComp);
```
My comments about `limitComp` above apply here as well.

peci1 replied:
done

azeey commented on test/integration/physics_system.cc:
```diff
@@ -812,6 +817,387 @@ TEST_F(PhysicsSystemFixture, ResetVelocityComponent)
+              }
+              if (iteration >= 30u && iteration < 40u)
+              {
+                if (iteration == 30u)
```
nit: Can this be simplified by using `_ecm.SetComponentData`?

peci1 replied:
good idea, implemented in all relevant places

azeey reviewed (changes requested):
Looks pretty good. I have a few comments that should be easy to address.

peci1 commented:
I addressed all review comments and added a new (de)serialization test for the newly added components.

I was thinking about removing the `Cmd` suffix from the components (as was requested in the ign-physics part), but here it seems to me the suffix makes sense. I can imagine that in the future, somebody will want to create `JointPositionLimits` read-only component (although the information is already accessible via the axis component).

azeey commented on test/integration/components.cc:
```diff
@@ -595,6 +598,108 @@ TEST_F(ComponentsTest, JointAxis)
+
+  std::istringstream istr("ignored");
+  components::JointEffortLimitsCmd comp3;
+  comp3.Deserialize(istr);
```
Thanks for adding these tests. I see expectations on the output of `Serialize`. Can you add expectations on the output of `Deserialize`?

peci1 replied:
Added

azeey reviewed (approved):
Approving with one minor comment.

chapulina commented:
Waiting until the end of September to release `ign-physics`

scpeters commented:
I recommend requiring version 2.5 of ignition-physics2

scpeters commented:
I think the docker cache is preventing new gzdev values in the Ubuntu builds:

~~~
Step 15/51 : RUN rm -fr /home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/gzdev
 ---> Using cache
 ---> 73265823fecb
Step 16/51 : RUN git clone --depth 1 https://github.com/ignition-tooling/gzdev -b master /home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/gzdev
 ---> Using cache
 ---> 5646bc68480b
Step 17/51 : RUN /home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/gzdev/gzdev.py repository enable --project=ignition-gazebo3 --force-linux-distro=bionic || ( git -C /home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/gzdev pull origin master &&     /home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/gzdev/gzdev.py repository enable --project=ignition-gazebo3 --force-linux-distro=bionic )
 ---> Using cache
 ---> a85cc93d1942
~~~

peci1 commented:
I increased the minimum required version.

scpeters commented:
looks like the docker cache has been refreshed, so the Ubuntu Jenkins jobs are passing, though Actions have some issues

azeey commented:
The failure in the Focal Github Action is because a different version of DART (6.9) is used there.  I see we've dealt with this before, but I'm not sure how it works:
https://github.com/ignitionrobotics/ign-gazebo/blob/632245103cb6e6b7849be5b34364f34a4e1218fd/test/integration/diff_drive_system.cc#L324-L328

peci1 commented:
maybe similar to https://github.com/ignitionrobotics/ign-physics/pull/267/commits/65397fb52de0d6d96792f11afaed0369bf4f61fa ?

azeey commented:
I added a check for DART 6.10.0 in https://github.com/ignitionrobotics/ign-gazebo/pull/847/commits/6b89ccb6f0fed576e0bae601ed7cb9ee4add86e2. I also added the `NOLINT` I asked you to remove from #869 because codecheck was failing 😅

azeey commented:
@osrf-jenkins run tests please
